README
mod_authnz_fido2 is an Apache 2.4 module for authorization by FIDO2 protocol,
without using an standalone WebAuthn server process.

All this is work in progress...

Build with: autoreconf -is && ./configure && make


debian/changelog
libapache2-mod-authnz-fido2 (0.1) unstable; urgency=medium

  * Initial release

 -- Roman Hodek <roman@hodek.net>  Sat, 21 Mar 2020 12:56:13 +0100


conf.c
/* mod_authn_fido2
 * Apache module for authentication with FIDO2 (WebAuthn)
 *
 * Roman Hodek <roman@hodek.net>
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>

#include <apr_strings.h>
#define APR_WANT_STRFUNC
#include <apr_want.h>
#include <ap_config.h>
#include <ap_provider.h>
#include <httpd.h>
#include <http_config.h>

#include "mod_authnz_fido2.h"


const command_rec authnz_fido2_cmds[] = {
	AP_INIT_TAKE1("AuthFIDO2RelyingPartyID", ap_set_string_slot,
				  (void *)APR_OFFSETOF(fido2_config_t, rpid_str),
				  OR_AUTHCFG,
				  "The (textual) relying party ID, usually the domain name of "
				  "the website or similar."),
	AP_INIT_TAKE1("AuthFIDO2UserFile", ap_set_file_slot,
				  (void *)APR_OFFSETOF(fido2_config_t, user_file),
				  OR_AUTHCFG,
				  "The filename of allowed users config."),
	AP_INIT_TAKE1("AuthFIDO2CookieName", ap_set_string_slot,
				  (void *)APR_OFFSETOF(fido2_config_t, cookie_name),
				  OR_AUTHCFG,
				  "The name of the session cookie used."),
	AP_INIT_TAKE1("AuthFIDO2Timeout", ap_set_int_slot,
				  (void *)APR_OFFSETOF(fido2_config_t, auth_timeout),
				  OR_AUTHCFG,
				  "Timeout (in seconds) for the authenticator at client side."),
	AP_INIT_TAKE1("AuthFIDO2TokenValidity", ap_set_int_slot,
				  (void *)APR_OFFSETOF(fido2_config_t, token_validity),
				  OR_AUTHCFG,
				  "Valid time of a token (in minutes)."),
	AP_INIT_TAKE1("AuthFIDO2TokenKeyLifetime", ap_set_int_slot,
				  (void *)APR_OFFSETOF(fido2_config_t, jwtkey_lifetime),
				  OR_AUTHCFG,
				  "Lifetime (in minutes) of the key for signing tokens."),
	AP_INIT_FLAG("AuthFIDO2OfferAllUsers", ap_set_flag_slot,
				 (void *)APR_OFFSETOF(fido2_config_t, offer_all_users),
				 OR_AUTHCFG,
				 "Offer all known users in allowedCredentials as "
				 "workaround if RPID-based authentication doesn't work."),
	AP_INIT_FLAG("AuthFIDO2RequireUserVerification", ap_set_flag_slot,
				 (void *)APR_OFFSETOF(fido2_config_t, require_UV),
				 OR_AUTHCFG,
				 "Request and require FIDO2 user verification flag."),
	{ NULL }
};

void *create_authnz_fido2_config(apr_pool_t *p, char *dirspec)
{
	fido2_config_t *conf = apr_pcalloc(p, sizeof(*conf));

	conf->rpid_str = NULL;
	memset(conf->rpid_hash, 0, sizeof(conf->rpid_hash));
	conf->user_file = NULL;
	conf->cookie_name = NULL;
	conf->offer_all_users = -1;
	conf->require_UV = -1;
	conf->auth_timeout = -1;
	conf->token_validity = -1;
	conf->jwtkey_lifetime = -1;
	return conf;
}

void *merge_authnz_fido2_config(apr_pool_t *p, void *base_conf, void *add_conf)
{
	fido2_config_t *base = base_conf;
	fido2_config_t *add  = add_conf;
	fido2_config_t *res  = apr_pcalloc(p, sizeof(*res));

#define merge_config(field,dflt)									\
	do {															\
		res->field = add->field >= 0 ? add->field : base->field;	\
		if (res->field < 0)											\
			res->field = dflt;										\
	} while(0)

	if (add->rpid_str) {
		res->rpid_str = add->rpid_str;
		sha256_str(res->rpid_str, res->rpid_hash);
	}
	else {
		res->rpid_str = base->rpid_str;
		memcpy(res->rpid_hash, base->rpid_hash, sizeof(res->rpid_hash));
	}
	res->user_file = add->user_file ?: base->user_file;
	res->cookie_name = add->cookie_name ?: base->cookie_name;

	merge_config(offer_all_users, DEFAULT_OFFER_ALL);
	merge_config(require_UV, DEFAULT_REQUIRE_UV);
	merge_config(auth_timeout, DEFAULT_AUTH_TIMEOUT);
	merge_config(token_validity, DEFAULT_TOKEN_VALID);
	merge_config(jwtkey_lifetime, DEFAULT_JWTKEY_LIFETIME);
	return res;
}



util.c
/* mod_authn_fido2
 * Apache module for authentication with FIDO2 (WebAuthn)
 *
 * Roman Hodek <roman@hodek.net>
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <apr_strings.h>
#include <apr_base64.h>
#include <httpd.h>
#include <http_core.h>
#include <http_log.h>
#include <fido.h>
#include <fido/es256.h>
#include <fido/rs256.h>
#include <fido/eddsa.h>
#include <openssl/evp.h>
#include <openssl/ec.h>
#include <openssl/ecdsa.h>
#include <openssl/pem.h>

#include "mod_authnz_fido2.h"



static fido2_user_t *getuser_byXXX(request_rec *req, fido2_config_t *conf,
								   const char *xxx, unsigned fno,
								   int (*callback)(const fido2_user_t *u))
{
	FILE *f;
	char line[1024];
	const int NFIELDS = 4;
	fido2_user_t *usr = NULL;

	if (!(f = fopen(conf->user_file, "r")))
		return NULL;
	while(fgets(line, sizeof(line), f)) {
		char *v[NFIELDS], *p;
		unsigned i;
		unsigned cntr;

		if (line[0] == '\n' || line[0] == '#')
			continue;

		for(p = line, i = 0; i < NFIELDS; ++i) {
			v[i] = p;
			if (!*(p += strcspn(p, ":")))
				goto next;
			*p++ = '\0';
		}			
		cntr = strtoul(p, NULL, 10);
		
		if (xxx && streq(v[fno], xxx)) {
			usr = apr_pcalloc(req->pool, sizeof(*usr));
			usr->name = apr_pstrdup(req->pool, v[0]);
			usr->credid = apr_pstrdup(req->pool, v[1]);
			usr->ktype  = apr_pstrdup(req->pool, v[2]);
			usr->pubkey = apr_pstrdup(req->pool, v[3]);
			usr->counter = cntr;
			break;
		}
		if (callback) {
			fido2_user_t u;
			u.name   = v[0];
			u.credid = v[1];
			u.ktype  = v[2];
			u.pubkey = v[3];
			u.counter = cntr;
			if (callback(&u))
				break;
		}
	  next:
		;
	}
	fclose(f);
	if (usr)
		return usr;

	errno = ESRCH;
	return NULL;
}

fido2_user_t *getuser_byname(request_rec *req, fido2_config_t *conf, const char *name)
{
	return getuser_byXXX(req, conf, name, 0, NULL);
}
fido2_user_t *getuser_bycredid(request_rec *req, fido2_config_t *conf, const char *credid)
{
	return getuser_byXXX(req, conf, credid, 1, NULL);
}

void for_all_users(request_rec *req, fido2_config_t *conf, int (*callback)(const fido2_user_t *u))
{
	getuser_byXXX(req, conf, NULL, 0, callback);
}

const char *get_rpid(request_rec *req, fido2_config_t *conf)
{
	return conf->rpid_str ?: req->hostname;
}

char *parse_cookie(request_rec *req, const char *cookiename)
{
	const char *cookie = (char*)apr_table_get(req->headers_in, "Cookie");
	char *str, *save, *rv = NULL;
	const char *p;
	
	if (!cookie)
		return NULL;

	/* need a writable copy for strtok() */
	str = alloca(strlen(cookie+1));
	strcpy(str, cookie);

	for(p = strtok_r(str, ";", &save); p; p = strtok_r(NULL, ";", &save)) {
		p += strspn(p, " ");
		if ((p = strprefix(p, cookiename)) && *p == '=') {
			rv = apr_pstrdup(req->pool, p+1);
			break;
		}
	}
	return rv;
}

void base64url2normal(char *str)
{
	char *p;
	for(p = str; *p; ++p)
		switch(*p) {
		  case '-': *p = '+'; break;
		  case '_': *p = '/'; break;
		}
}

void remove_slashes(char *str)
{
	char *p = str+strlen(str)-1;
	while(p > str && *p == '/')
		*p-- = '\0';
}

int get_ktype(const char *str)
{
	if (streq(str, "es256"))
		return COSE_ES256;
	else if (streq(str, "rs256"))
		return COSE_RS256;
	else if (streq(str, "eddsa"))
		return COSE_EDDSA;
	else
		return 0;
}

void *get_pubkey(int ktype, const char *kdata)
{
	void *ret = NULL;
	unsigned pkey_len = apr_base64_decode_len(kdata);
	uint8_t pkey_data[pkey_len];
	const uint8_t *pkey = pkey_data;
	pkey_len = apr_base64_decode(pkey_data, kdata);

	switch(ktype) {
	  case COSE_ES256: {
		EC_KEY *ec = d2i_EC_PUBKEY(NULL, &pkey, pkey_len);
		es256_pk_t *es256_pk;
		if (!ec)
			return NULL;
		if (!(es256_pk = es256_pk_new()) || es256_pk_from_EC_KEY(es256_pk, ec)) {
			EC_KEY_free(ec);
			return NULL;
		}
		ret = es256_pk;
		EC_KEY_free(ec);
		break;
	  }
	  case COSE_RS256: {
		RSA *rk = d2i_RSA_PUBKEY(NULL, &pkey, pkey_len);
		rs256_pk_t *rs256_pk;
		if (!rk)
			return NULL;
		if (!(rs256_pk = rs256_pk_new()) || rs256_pk_from_RSA(rs256_pk, rk)) {
			RSA_free(rk);
			return NULL;
		}
		ret = rs256_pk;
		RSA_free(rk);
	  }
	  case COSE_EDDSA: {
		eddsa_pk_t *eddsa_pk;
		if (!(eddsa_pk = eddsa_pk_new()) ||
			eddsa_pk_from_ptr(eddsa_pk, pkey, pkey_len))
			return NULL;
		ret = eddsa_pk;
	  }
	  default:
		return NULL;
	}
	return ret;
}

void free_pubkey(int ktype, void *pk)
{
	switch(ktype) {
	  case COSE_ES256: {
		es256_pk_t *es256_pk = pk;
		es256_pk_free(&es256_pk);
		break;
	  }
	  case COSE_RS256: {
		rs256_pk_t *rs256_pk = pk;
		rs256_pk_free(&rs256_pk);
		break;
	  }
	  case COSE_EDDSA: {
		eddsa_pk_t *eddsa_pk;
		eddsa_pk_free(&eddsa_pk);
		break;
	  }
	}
}

int sha256(const uint8_t *in, size_t inlen, uint8_t *out)
{
	EVP_MD_CTX *ctx;
	int rv = -1;
	unsigned reallen;

	if (!(ctx = EVP_MD_CTX_create()))
		return -1;

	if (!EVP_DigestInit_ex(ctx, EVP_sha256(), NULL))
		goto out;
	if (!EVP_DigestUpdate(ctx, in, inlen))
		goto out;
	if (!EVP_DigestFinal_ex(ctx, out, &reallen))
		goto out;
	if (reallen != SHA256_LEN)
		goto out;
	rv = 0;

  out:
	EVP_MD_CTX_destroy(ctx);
	return rv;
}

int sha256_2buf(const uint8_t *in1, size_t in1len, const uint8_t *in2, size_t in2len, uint8_t *out)
{
	EVP_MD_CTX *ctx;
	int rv = -1;
	unsigned reallen;

	if (!(ctx = EVP_MD_CTX_create()))
		return -1;

	if (!EVP_DigestInit_ex(ctx, EVP_sha256(), NULL))
		goto out;
	if (!EVP_DigestUpdate(ctx, in1, in1len))
		goto out;
	if (!EVP_DigestUpdate(ctx, in2, in2len))
		goto out;
	if (!EVP_DigestFinal_ex(ctx, out, &reallen))
		goto out;
	if (reallen != SHA256_LEN)
		goto out;
	rv = 0;

  out:
	EVP_MD_CTX_destroy(ctx);
	return rv;
}

int sha256_3buf(const uint8_t *in1, size_t in1len, const uint8_t *in2, size_t in2len, const uint8_t *in3, size_t in3len, uint8_t *out)
{
	EVP_MD_CTX *ctx;
	int rv = -1;
	unsigned reallen;

	if (!(ctx = EVP_MD_CTX_create()))
		return -1;

	if (!EVP_DigestInit_ex(ctx, EVP_sha256(), NULL))
		goto out;
	if (!EVP_DigestUpdate(ctx, in1, in1len))
		goto out;
	if (!EVP_DigestUpdate(ctx, in2, in2len))
		goto out;
	if (!EVP_DigestUpdate(ctx, in3, in3len))
		goto out;
	if (!EVP_DigestFinal_ex(ctx, out, &reallen))
		goto out;
	if (reallen != SHA256_LEN)
		goto out;
	rv = 0;

  out:
	EVP_MD_CTX_destroy(ctx);
	return rv;
}

int sha256_str(const char *str, uint8_t *out)
{
	return sha256(str, strlen(str), out);
}

void log_bytearray(request_rec *req, const char *prefix, uint8_t *data, size_t len)
{
	char buf[16*3+1+1], *p = buf;
	unsigned i;

	for(i = 0; i < len; ++i) {
		if ((i % 16) == 0)
			*(p = buf) = '\0';
		p += sprintf(p, "%02x ", data[i]);
		if ((i % 16) == 7)
			strcpy(p++, " ");
		if ((i % 16) == 15)
			debug("%s %04x: %s", prefix, i&~0xf, buf);
						  
	}
	if (len % 16)
		debug("%s %04x: %s", prefix, i&~0xf, buf);
}


mod_authnz_fido2.c
/* mod_authn_fido2
 * Apache module for authentication with FIDO2 (WebAuthn)
 *
 * Roman Hodek <roman@hodek.net>
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <apr_strings.h>
#define APR_WANT_STRFUNC
#include <apr_want.h>
#include <ap_config.h>
#include <ap_provider.h>
#include <apr_base64.h>
#include <httpd.h>
#include <http_config.h>
#include <http_core.h>
#include <http_log.h>
#include <http_protocol.h>
#include <http_request.h>
#include <mod_auth.h>
#include <mod_ssl.h>
#include <openssl/crypto.h>
#include <openssl/rand.h>
#include <openssl/evp.h>
#include <jansson.h>
#include <fido.h>

#include "mod_authnz_fido2.h"


#define NAME		"mod_authn_fido2"
#define VERSION 	"0.1"

typedef struct {
	uint8_t *sessiondata;
	size_t  sessiondata_len;
	char *credid;
	uint8_t *authdata;
	size_t  authdata_len;
	uint8_t *signature;
	size_t  signature_len;
	uint8_t *cdata;
	size_t  cdata_len;
	/* substructure 'clientData' */
	uint8_t *cd_challenge;
	size_t  cd_challenge_len;
	const char *cd_origin;
	const char *cd_type;
} assert_resp_t;

typedef struct {
	uint8_t rp_id_hash[32];
	uint8_t flags;
	uint8_t counter[4];
} raw_authenticator_data_t;

typedef struct {
	uint8_t rp_id_hash[SHA256_LEN];
	unsigned flags;
	unsigned counter;
} authenticator_data_t;

static uint8_t *chmackey;

module AP_MODULE_DECLARE_DATA authnz_fido2_module;

static const char *login_html = "\
<html>\n\
<head>\n\
  <title>mod_authn_fido2 login</title>\n\
</head>\n\
<body>\n\
  <h3>Apache mod_authn_fido2 Login</h3>\n\
  <p>This requires a browser supporting the WebAuthn API</p>\n\
  <p>Trigger your authenticator device now...</p>\n\
  <script>\n\
	function enc(arr) {\n\
		return btoa(String.fromCharCode.apply\n\
					(null, new Uint8Array(arr)));\n\
	}\n\
	function dec(str) {\n\
		return Uint8Array.from(Array.prototype.map.call(atob(str),\n\
			   function(x) { return x.charCodeAt(0); }));\n\
	}\n\
    navigator.credentials.get(%s)\n\
    .then((ar) => {\n\
      return fetch('%s', {\n\
        method: 'POST',\n\
        headers: {'Content-Type': 'application/json'},\n\
        body: '{ '+\n\
		  '\"sessiondata\":\"%s\",'+\n\
		  '\"credentialId\":\"'+enc(ar.rawId)+'\", '+\n\
		  '\"authenticatorData\":\"\'+enc(ar.response.authenticatorData)+'\", '+\n\
		  '\"clientDataJSON\":\"'+enc(ar.response.clientDataJSON)+'\", '+\n\
		  '\"signature\":\"'+enc(ar.response.signature)+'\" }'\n\
      })})\n\
    .then((res) => {\n\
	  if (res.ok) { return res.text(); }\n\
	  else { alert('Authentication failed.'); throw 'failed'; }\n\
    }, (reason) => {\n\
      alert(reason); throw 'failed';\n\
    })\n\
    .then((token) => {\n\
      document.cookie='%s='+token;\n\
      window.location.reload();\n\
    }, (reason) => {});\n\
  </script>\n\
</body></html>\n\
";
	

/* ---------------------------------------------------------------------- */


static int send_webauthn_code(request_rec *req, fido2_config_t *conf)
{
	uint8_t challenge[CHALLENGE_LEN];
	uint8_t hash[CHMACSTAMP_LEN+SHA256_LEN];
	time_t  stamp;
	char sessiondata_str[SHA256_LEN*2];
	char challenge_str[CHALLENGE_LEN*2];
	char allowed_ids[1024] = "";
	const char *cookie_name = conf->cookie_name ?: "modfido2session";

	/* A simple HMAC with a short key (128b currently) and a timestamp
	 * should be enough to avoid challenge trickeries.
	 * The layout for hashing and sessiondata is:
	 *
	 *    chall (32B)  time (4B)  hmackey (16B)
	 *    |            |       |              |
	 *    +------------------------hashed-----+ => hash (32B)
	 *                 |       |                   |        |
	 *                 +-------+  =>  base64  <=   +--------+
	 *                             sessiondata
	 */
	RAND_bytes(challenge, CHALLENGE_LEN);
	time(&stamp);
	memcpy(hash, &stamp, CHMACSTAMP_LEN);
	sha256_3buf(challenge, CHALLENGE_LEN,
				(uint8_t*)&stamp, CHMACSTAMP_LEN,
				chmackey, CHMACKEY_LEN,
				hash+CHMACSTAMP_LEN);
	apr_base64_encode(sessiondata_str, hash, CHMACSTAMP_LEN+SHA256_LEN);
	apr_base64_encode(challenge_str, challenge, CHALLENGE_LEN);
	memset(challenge, 0, sizeof(challenge));

	if (conf->offer_all_users) {
		char *idstr = allowed_ids;
		size_t idlen = sizeof(allowed_ids), l;
		int first = 1;

		int add_credid(const fido2_user_t *u) {
			l = snprintf(idstr, idlen, "%s{'type':'public-key','id':dec('%s')}",
						 first ? "" : ", ", u->credid);
			idstr += l; idlen -= l;
			first = 0;
			return 0;
		}

		l = snprintf(idstr, idlen, ", 'allowCredentials':[");
		idstr += l; idlen -= l;
		for_all_users(req, conf, add_credid);
		l = snprintf(idstr, idlen, "]");
		idstr += l; idlen -= l;
	}

	char *obj_str = apr_psprintf(
		req->pool,
		"{ 'publicKey': { "
		"'rpId': '%s', "
		"'userVerification': '%s', "
		"'timeout': %d, "
		"'challenge': dec('%s')"
		"%s"
		"} }",
		get_rpid(req, conf),
		conf->require_UV ? "required" : "discouraged",
		conf->auth_timeout * 1000,
		challenge_str,
		allowed_ids
		);

	req->user = "nobody";
	ap_set_content_type(req, "text/html; charset=US-ASCII");
	apr_table_setn(req->headers_out, "Cache-Control", "no-cache");
	ap_rprintf(req, login_html, obj_str, req->uri, sessiondata_str, cookie_name);

	memset(sessiondata_str, 0, sizeof(sessiondata_str));
	memset(challenge_str, 0, sizeof(challenge_str));
	return DONE;
}

static char *apr_get_postdata(request_rec *req, apr_size_t *len)
{
	char *data = NULL;

	*len = 0;
	ap_setup_client_block(req, REQUEST_CHUNKED_ERROR);
	
	if (ap_should_client_block(req) == 1) {
		apr_off_t n, pos = 0;
		char buf[1024];

		*len = req->remaining;
		data = apr_pcalloc(req->pool, *len);
		while((n = ap_get_client_block(req, buf, sizeof(buf))) > 0) {
			if (pos+n > *len)
				n = *len-pos;
			memcpy(data+pos, buf, n);
			pos += n;				   
		}
	}
	//debug("rcv POST data: %s", data);
	return data;
}

static const char *parse_assertation(request_rec *req, uint8_t *buf, size_t len, assert_resp_t *resp)
{
	json_error_t jerr;
	json_t *top = json_loadb(buf, len, 0, &jerr), *jcdata, *node;

#define get_str_common(obj,name)							\
		const char *__val;									\
		if (!(node = json_object_get(obj, name)))			\
			return "assertation."name" not found";			\
		if (!json_is_string(node))							\
			return "assertation."name" is not a string";	\
		__val = json_string_value(node);
#define get_dec_base64(val,field)							\
		field##_len = apr_base64_decode_len(val);			\
		if (!(field = apr_pcalloc(req->pool, field##_len)))	\
			return "out of memory";							\
		field##_len = apr_base64_decode(field, val);

#define get_str_attr(obj,name,field)			\
	do {										\
		get_str_common(obj,name);				\
		field = apr_pstrdup(req->pool, __val);	\
	} while(0)
#define get_base64_attr(obj,name,field)			\
	do {										\
		get_str_common(obj,name);				\
		get_dec_base64(__val,field);			\
	} while(0)
#define get_base64url_attr(obj,name,field)		\
	do {										\
		get_str_common(obj,name);				\
		char *__str = alloca(strlen(__val)+1);	\
		strcpy(__str, __val);					\
		base64url2normal(__str);				\
		get_dec_base64(__str,field);			\
	} while(0)

	if (!top || !json_is_object(top))
		return apr_pstrdup(req->pool, jerr.text);
	get_base64_attr(top, "sessiondata", resp->sessiondata);
	get_str_attr(top, "credentialId", resp->credid);
	get_base64_attr(top, "authenticatorData", resp->authdata);
	get_base64_attr(top, "signature", resp->signature);

	get_base64_attr(top, "clientDataJSON", resp->cdata);
	jcdata = json_loadb(resp->cdata, resp->cdata_len,
						JSON_DISABLE_EOF_CHECK, &jerr);
	if (!jcdata || !json_is_object(jcdata))
		return apr_pstrdup(req->pool, jerr.text);

	get_base64url_attr(jcdata, "challenge", resp->cd_challenge);
	get_str_attr(jcdata, "origin", resp->cd_origin);
	get_str_attr(jcdata, "type", resp->cd_type);

	return NULL;
}

static void decode_authenticator_data(authenticator_data_t *out,
									  raw_authenticator_data_t *in)
{
	memcpy(out->rp_id_hash, in->rp_id_hash, sizeof(out->rp_id_hash));
	out->flags = in->flags;
	out->counter = in->counter[3] |
				   (in->counter[2]<<8) |
				   (in->counter[1]<<16) |
				   (in->counter[0]<<24);
}

static int process_webauthn_reply(request_rec *req, fido2_config_t *conf)
{
	const char *ctype = apr_table_get(req->headers_in, "Content-Type");
	const char *p;
	char *postdata;
	apr_off_t postlen;
	assert_resp_t ar;
	const char *errstr;
	uint8_t hash[SHA256_LEN];
	apr_time_t now = apr_time_sec(apr_time_now());
	time_t stamp;
	authenticator_data_t authdata;
	fido2_user_t *uent;
	int ktype;
	void *pk;
	fido_assert_t *ass;
	int r;
	char *token_str;

	if (!ctype || strcmp(ctype, "application/json") != 0) {
		error("content-type must be application/json");
		return HTTP_UNSUPPORTED_MEDIA_TYPE;
	}

	postdata = apr_get_postdata(req, &postlen);
	if ((errstr = parse_assertation(req, postdata, postlen, &ar))) {
		debug("parse error in posted assertation data: %s", errstr);
		return HTTP_INTERNAL_SERVER_ERROR;
	}

	/* NB: The WebAuthn standard says in 7.2.1 "If the allowCredentials option
	 * was given, verify that credential.id identifies one of those that were
	 * listed in allowCredentials." This is deliberately not implemented here,
	 * as we either give no credentials in that list, or all of them; so this
	 * check is pointless. */

	if (strcmp(ar.cd_type, "webauthn.get") != 0) {
		error("wrong cd_type");
		return HTTP_BAD_REQUEST;
	}
	/* check correct origin (should be https:// + our name or localhost, where
	 * also http is allowed) */
	if (!(streq(ar.cd_origin, "http://localhost") ||
		  strprefix(ar.cd_origin, "http://localhost:") ||
		  ((p = strprefix(ar.cd_origin, "https://")) &&
		   streq(p, req->hostname)))) {
		warn("wrong origin (%s vs. expected %s)",
			 ar.cd_origin+strlen("https://"), req->hostname);
		return HTTP_BAD_REQUEST;
	}

	/* check challenge HMAC */
	if (ar.cd_challenge_len != CHALLENGE_LEN) {
		error("bad challenge size in response");
		return HTTP_BAD_REQUEST;
	}
	if (ar.sessiondata_len != CHMACSTAMP_LEN+SHA256_LEN) {
		error("bad sessiondata size in response");
		return HTTP_BAD_REQUEST;
	}
	sha256_3buf(ar.cd_challenge, CHALLENGE_LEN,
				ar.sessiondata, CHMACSTAMP_LEN,
				chmackey, CHMACKEY_LEN, hash);
	if (memcmp(ar.sessiondata+CHMACSTAMP_LEN, hash, SHA256_LEN) != 0) {
		error("challenge HMAC failure in reply");
		return HTTP_BAD_REQUEST;
	}
	memcpy(&stamp, ar.sessiondata, CHMACSTAMP_LEN); // copy for alignment
	/* The timestamp in sessiondata (= when challenge has been generated)
	 * must be not older then the authentication timeout plus a generous 15s
	 * for network delays. It also must not be in future.
	 */
	//debug("sessiondata tstamp=%ld age=%ld", stamp, now-stamp);
	if (stamp > now || stamp < now - conf->auth_timeout - 15) {
		error("challenge in reply outside allowed time range");
		return HTTP_BAD_REQUEST;
	}

	if (ar.authdata_len != sizeof(raw_authenticator_data_t)) {
		warn("bad authenticator_data length (%ld vs. expected %ld",
			 ar.authdata_len, sizeof(raw_authenticator_data_t));
		return HTTP_BAD_REQUEST;
	}
	decode_authenticator_data(&authdata,
							  (raw_authenticator_data_t*)ar.authdata);

	if (!(uent = getuser_bycredid(req, conf, ar.credid))) {
		error("unknown credential ID %s", ar.credid);
		return HTTP_BAD_REQUEST;
	}
	ktype = get_ktype(uent->ktype);
	if (!ktype) {
		error("unsupported key type %s for CredID %s", uent->ktype, ar.credid);
		return HTTP_INTERNAL_SERVER_ERROR;
	}
	debug("credID=%s -> user %s ktype %s", ar.credid, uent->name, uent->ktype);

	if (!(pk = get_pubkey(ktype, uent->pubkey))) {
		error("failed to get %s key for CredID %s", uent->ktype, ar.credid);
		return HTTP_INTERNAL_SERVER_ERROR;
	}

	if (!(ass = fido_assert_new()) || fido_assert_set_count(ass, 1)) {
		error("cannot create assertation");
		return HTTP_INTERNAL_SERVER_ERROR;
	}
	sha256(ar.cdata, ar.cdata_len, hash);
	if ((r = fido_assert_set_rp(ass, get_rpid(req, conf))) ||
		(r = fido_assert_set_clientdata_hash(ass, hash, SHA256_LEN)) ||
		(r = fido_assert_set_authdata_raw(ass, 0, ar.authdata, ar.authdata_len)) ||
		(r = fido_assert_set_sig(ass, 0, ar.signature, ar.signature_len)) ||
		(r = fido_assert_set_up(ass, FIDO_OPT_TRUE)) ||
		(r = (conf->require_UV ? fido_assert_set_uv(ass, FIDO_OPT_TRUE):0))) {
		error("cannot set data into assertation object: %s", fido_strerr(r));
		fido_assert_free(ass);
		return HTTP_INTERNAL_SERVER_ERROR;
	}

	r = fido_assert_verify(ass, 0, ktype, pk);
	free_pubkey(ktype, pk);
	fido_assert_free(ass);

	if (r) {
		error("FIDO2 verification error %d: %s", r, fido_strerr(r));
		return HTTP_INTERNAL_SERVER_ERROR;
	}

	if (!(token_str = create_token(req, conf, uent)))
		return HTTP_INTERNAL_SERVER_ERROR;
	ap_set_content_type(req, "application/json");
	ap_rwrite(token_str, strlen(token_str), req);

	req->user = (char*)uent->name;
	return DONE;
}

static int fido2_handler(request_rec *req)
{
	fido2_config_t *conf =
		ap_get_module_config(req->per_dir_config, &authnz_fido2_module);
	const char *cookie_name = conf->cookie_name ?: "modfido2session";
    const char *auth_type = ap_auth_type(req);
    const char *auth_name = ap_auth_name(req);
	char *session;

	debug("called, auth_type='%s'", auth_type);
	if (!auth_type || strcasecmp(auth_type, "fido2") != 0)
		return DECLINED;
	// XXX still needed?
	if (!auth_name) {
		error("no AuthName defined");
		return HTTP_INTERNAL_SERVER_ERROR;
	}
	
	/* Check for session cookie with access token */
	session = parse_cookie(req, cookie_name);
	if (session && *session) {
		char *user;
		int rv;

		if ((rv = check_token(req, conf, session, &user)) == OK)
			req->user = user;
		return rv;
	}
	else {
		/* If there's no Authorization: header, redirect GET to login page
		 * and process reply in POST handler
		 */
		if (strcmp(req->method, "GET") == 0) {
			debug("no session, GET, reply with login HTML");
			return send_webauthn_code(req, conf);
		}
		else if (strcmp(req->method, "POST") == 0) {
			debug("no session, POST, process auth data");
			return process_webauthn_reply(req, conf);
		}
		else {
			debug("other method: %s", req->method);
			return HTTP_INTERNAL_SERVER_ERROR;
		}
	}
   
    return DECLINED;
}

/* ---------------------------------------------------------------------- */
/* registration															  */

static int init_mod_fido2(apr_pool_t *pconf, apr_pool_t *plog,
						  apr_pool_t *ptemp, server_rec *s)
{
	request_rec *req = (request_rec*)s; // just for error/warn macros
	uint8_t *jwtkey_mem;
	
    ap_add_version_component(pconf, NAME "/" VERSION);

	/* Use OpenSSL secure heap for our keys -- protect them as good as
	 * possible. */
	if (!CRYPTO_secure_malloc_initialized()) {
		switch(CRYPTO_secure_malloc_init(8*1024, 16)) {
		  case 0:
			error("secure_malloc_init failed");
			return HTTP_INTERNAL_SERVER_ERROR;
		  case 2:
			warn("(no OpenSSL secured memory available)");
			break;
		}
	}
	if (!(jwtkey_mem = OPENSSL_secure_malloc(JWTKEY_LEN*JWTKEY_NUM)) ||
		!(chmackey = OPENSSL_secure_malloc(CHMACKEY_LEN))) {
		error("secure_malloc failed");
		return HTTP_INTERNAL_SERVER_ERROR;
	}
	RAND_bytes(chmackey, CHMACKEY_LEN);
	jwtkey_init(jwtkey_mem);

    return OK;
}

static void register_hooks(apr_pool_t *p)
{
    ap_hook_check_authn(fido2_handler, NULL, NULL,
						APR_HOOK_MIDDLE, AP_AUTH_INTERNAL_PER_URI);
	ap_hook_post_config(init_mod_fido2, NULL, NULL, APR_HOOK_MIDDLE);
}

AP_DECLARE_MODULE(authnz_fido2) = {
	STANDARD20_MODULE_STUFF,
	create_authnz_fido2_config,
	merge_authnz_fido2_config,
	NULL, /* create per-server config structure */
	NULL, /* merge per-server config structure */
	authnz_fido2_cmds,
	register_hooks
};
